SnowplowTracker/SnowplowTracker/Tracker.cs

```diff
@@ -0,0 +1,339 @@
+		}
+
+		// --- Tracker Control Methods
+
+		/// <summary>
+		/// Pauses the event tracking.
+		/// </summary>
+		public void PauseEventTracking() {
+			dataCollection = false;
+			if (session != null) {
+				session.StopSessionChecker();
+				emitter.Stop();
+			}
+		}
+
+		/// <summary>
+		/// Resumes the event tracking.
+		/// </summary>
+		public void ResumeEventTracking() {
+			dataCollection = true;
+			if (session != null) {
+				session.StartSessionChecker();
+				emitter.Start();
+			}
+		}
+
+		/// <summary>
+		/// Sets the tracker to synchronous operation.
+		/// </summary>
+		public void SetSynchronous() {
+			synchronous = true;
+			this.emitter.synchronous = true;
+		}
+
+		/// <summary>
+		/// Sets the tracker to asynchronous operation.
+		/// </summary>
+		public void SetAsynchronous() {
+			synchronous = false;
+			this.emitter.synchronous = false;
+		}
+
+		// --- Event Tracking
+
+		/// <summary>
+		/// Track the specified event.
+		/// - Checks whether dataCollection is on
+		/// - Checks whether to use synchronous sending or not
+		/// </summary>
+		/// <param name="newEvent">New event to track.</param>
+		public void Track(IEvent newEvent) {
+			if (!dataCollection) {
+				return;
+			}
+			if (synchronous) {
+				ProcessEvent (newEvent);
+			} else {
+				ThreadPool.QueueUserWorkItem (new WaitCallback ( delegate(object t) {
+					ProcessEvent (newEvent);
+				}));
+			}
+		}
+
+		/// <summary>
+		/// Processes the event.
+		/// </summary>
+		/// <param name="newEvent">New event.</param>
+		private void ProcessEvent(IEvent newEvent) {
+			List<IContext> contexts = newEvent.GetContexts ();
+			string eventId = newEvent.GetEventId();
+			Type eType = newEvent.GetType ();
+			
+			if (eType == typeof(PageView) || eType == typeof(Structured)) {
+				AddTrackerPayload ((TrackerPayload)newEvent.GetPayload (), contexts, eventId);
+			} else if (eType == typeof(EcommerceTransaction)) {
+				AddTrackerPayload ((TrackerPayload)newEvent.GetPayload (), contexts, eventId);
+				EcommerceTransaction ecommerceTransaction = (EcommerceTransaction) newEvent;
+				foreach (EcommerceTransactionItem item in ecommerceTransaction.GetItems()) {
+					item.SetTimestamp(ecommerceTransaction.GetTimestamp());
+					AddTrackerPayload ((TrackerPayload)item.GetPayload (), item.GetContexts (), item.GetEventId ());
+				}
+			} else if (eType == typeof(Unstructured)) {
+				Unstructured unstruct = (Unstructured)newEvent;
+				unstruct.SetBase64Encode(this.base64Encoded);
+				AddTrackerPayload ((TrackerPayload)unstruct.GetPayload (), contexts, eventId);
+			} else if (eType == typeof(Timing) || eType == typeof(ScreenView)) {
+				this.ProcessEvent(new Unstructured()
+				           .SetEventData((SelfDescribingJson)newEvent.GetPayload())
+				           .SetCustomContext(newEvent.GetContexts())
+				           .SetTimestamp(newEvent.GetTimestamp())
+				           .SetEventId(newEvent.GetEventId())
+				           .Build());
+			}
+		}
```

Comment:
Rather than manually checking the type of each event, is it possible to make the logic which depends on the event type part of the ITracker interface, possibly as a method returning a sequence of Payloads? I don't know whether this would work easily.